cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,132 @@
+package io.strimzi.operator.cluster.operator.resource;
+
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.strimzi.operator.common.VertxUtil;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.QuorumInfo;
+
+import static java.lang.Math.ceil;
+
+/**
+ * Provides a method that determines whether it's safe to restart a KRaft controller
+ */
+class KafkaQuorumCheck {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaQuorumCheck.class.getName());
+    private final Reconciliation reconciliation;
+    private final Admin admin;
+    private final Vertx vertx;
+    private final long controllerQuorumFetchTimeoutMs;
+
+    protected KafkaQuorumCheck(Reconciliation reconciliation, Admin ac, Vertx vertx, long controllerQuorumFetchTimeoutMs) {
+        this.reconciliation = reconciliation;
+        this.admin = ac;
+        this.vertx = vertx;
+        this.controllerQuorumFetchTimeoutMs = controllerQuorumFetchTimeoutMs;
+    }
+
+    /**
+     * Returns future that completes with true if the given controller can be rolled based on the quorum state. Quorum is considered
+     * healthy if the majority of controllers have caught up with the quorum leader within the controller.quorum.fetch.timeout.ms.
+     */
+    Future<Boolean> canRollController(int nodeId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller pod {} can be rolled", nodeId);
+        return describeMetadataQuorum().map(info -> {
+            boolean canRoll = isQuorumHealthyWithoutNode(nodeId, info);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Not restarting controller pod {}. Restart would affect the quorum health", nodeId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart controller pod {}", nodeId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the controller quorum leader id");
+        return describeMetadataQuorum().map(QuorumInfo::leaderId).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the controller quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns true if the majority of the controllers' lastCaughtUpTimestamps (not including the given node) are within
+     * the controller.quorum.fetch.timeout.ms based on the given quorum info.
+     **/
+    private boolean isQuorumHealthyWithoutNode(int nodeId, QuorumInfo info) {
```

Comment:
maybe also changing the name of the variable to make it clearer ... something like `nodeToRollId` or `nodeIdToRoll` or anything else similar?